tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -68,258 +94,410 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
+        let mut test_count_per_package = HashMap::new();
+
+        for (package_name, package_tests) in package_test_results {
+            let package_tests = package_tests?;
+            test_count_per_package.insert(package_name, package_tests.len());
+            tests.extend(package_tests);
+        }
+
+        // If there were no compile errors we'll run them all. Here we show how many there are per package.
+        for (package_name, test_count) in &test_count_per_package {
+            let plural = if *test_count == 1 { "" } else { "s" };
+            println!("[{package_name}] Running {test_count} test function{plural}");
+        }
+
+        // Now run all tests in parallel, but show output for each package sequentially
+        let test_reports = self.run_all_tests(tests, &test_count_per_package);
+
+        if test_reports.is_empty() {
+            match &self.pattern {
+                FunctionNameMatch::Exact(pattern) => {
+                    return Err(CliError::Generic(format!(
+                        "Found 0 tests matching input '{pattern}'.",
+                    )))
+                }
+                FunctionNameMatch::Contains(pattern) => {
+                    return Err(CliError::Generic(
+                        format!("Found 0 tests containing '{pattern}'.",),
+                    ))
+                }
+                // If we are running all tests in a crate, having none is not an error
+                FunctionNameMatch::Anything => {}
+            };
+        }
+
+        if test_reports.iter().any(|test_result| test_result.status.failed()) {
+            Err(CliError::Generic(String::new()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn run_all_tests(
+        &self,
+        tests: Vec<Test<'a>>,
+        test_count_per_package: &HashMap<String, usize>,
+    ) -> Vec<TestResult> {
+        // Here we'll gather all test reports from all packages
+        let mut test_reports = Vec::new();
+
+        let (sender, receiver) = mpsc::channel();
+        let iter = &Mutex::new(tests.into_iter());
+        thread::scope(|scope| {
+            // Start worker threads
+            for _ in 0..self.num_threads {
+                // Clone sender so it's dropped once the thread finishes
+                let thread_sender = sender.clone();
+                thread::Builder::new()
+                    // Specify a larger-than-default stack size to prevent overflowing stack in large programs.
+                    // (the default is 2MB)
+                    .stack_size(STACK_SIZE)
+                    .spawn_scoped(scope, move || loop {
+                        // Get next test to process from the iterator.
+                        let Some(test) = iter.lock().unwrap().next() else {
+                            break;
+                        };
+
+                        let time_before_test = std::time::Instant::now();
+                        let (status, output) = (test.runner)();
```

Comment:
Should we use [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) to make sure no assertion in mocks or pedantic solvers bubble up?

Reply:
What I'd like to understand is that if somebody writes a test that hits for example a panic [like this one](https://github.com/noir-lang/noir/blob/4b0df947a3e74b633412743ff286954305f39938/acvm-repo/blackbox_solver/src/bigint.rs#L71-L72) by using some invalid value in a test, is that going to crash `nargo` itself, and if it does, is it going to tell before it happens which test was to blame?

Reply:
I just tried it. It seems that if there's a panic inside `run_test` then that panic is printed(it could happen in multiple threads) and then the program aborts.

If we use `catch_unwind` the program continues, but we can at least know the tests that failed because of an unexpected errors. So I'll make this change 👍